vpr/src/place/net_cost_handler.cpp

```diff
@@ -448,7 +453,88 @@ static double wirelength_crossing_count(size_t fanout);
 static void set_bb_delta_cost(const int num_affected_nets, double& bb_delta_c);
 
 /******************************* End of Function definitions ************************************/
+// Initialize the ts vectors
+void BB2D3DControlFlow::init(size_t num_nets, bool cube_bb_in) {
+    const int num_layers = g_vpr_ctx.device().grid.get_num_layers();
+
+    cube_bb = cube_bb_in;
+    if (cube_bb) {
+        ts_info.ts_bb_edge_new.resize(num_nets, t_bb());
+        ts_info.ts_bb_coord_new.resize(num_nets, t_bb());
+    } else {
+        VTR_ASSERT_SAFE(!cube_bb);
+        ts_info.layer_ts_bb_edge_new.resize(num_nets, std::vector<t_2D_bb>(num_layers, t_2D_bb()));
+        ts_info.layer_ts_bb_coord_new.resize(num_nets, std::vector<t_2D_bb>(num_layers, t_2D_bb()));
+    }
+
+    /*This initialize the whole matrix to OPEN which is an invalid value*/
+    ts_info.ts_layer_sink_pin_count.resize({num_nets, size_t(num_layers)}, OPEN);
 
+    ts_info.ts_nets_to_update.resize(num_nets, ClusterNetId::INVALID());
+}
+
+void BB2D3DControlFlow::get_non_updatable_bb(const ClusterNetId& net) {
+    if(cube_bb)
+        ::get_non_updatable_bb(net,
+                               ts_info.ts_bb_coord_new[net],
+                               ts_info.ts_layer_sink_pin_count[size_t(net)]);
+    else
+        ::get_non_updatable_layer_bb(net,
+                                   ts_info.layer_ts_bb_coord_new[net],
+                                   ts_info.ts_layer_sink_pin_count[size_t(net)]);
+}
+
+bool BB2D3DControlFlow::is_driver(const t_physical_tile_type_ptr& blk_type, const ClusterPinId& blk_pin) {
+    auto& cluster_ctx = g_vpr_ctx.clustering();
+    if(cube_bb)
+        return cluster_ctx.clb_nlist.pin_type(blk_pin) == PinType::DRIVER;
+    else
```

Comment:
I think we can do the if part all the time and it is simpler.  If so, we can just get rid of the routine.